### ISSUE TITLE
'Failed to emit module' error when compiling usages of a nested tasklike type

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#16493]
Roslyn doesn't detect errors in the following code snippet but is unable to compile it. Fails with 'failed to emit module' error in VS2017 RC.

```cs
public class Program
{
    public static void Main() { }
}

public class TypeParametersProvider<T>
{
    public async ValueTask Method()
    {
        await Task.Delay(5);
        return;
    }

    [AsyncMethodBuilder(typeof(AsyncValueTaskMethodBuilder))]
    public struct ValueTask
    {
    }
}

public class AsyncValueTaskMethodBuilder
{
    public TypeParametersProvider<int>.ValueTask Task { get => default(TypeParametersProvider<int>.ValueTask); }
    public void AwaitOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter awaiter, ref TStateMachine stateMachine) where TAwaiter : INotifyCompletion where TStateMachine : IAsyncStateMachine { }
    public void AwaitUnsafeOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter awaiter, ref TStateMachine stateMachine) where TAwaiter : ICriticalNotifyCompletion where TStateMachine : IAsyncStateMachine { }
    public static AsyncValueTaskMethodBuilder Create() { return default(AsyncValueTaskMethodBuilder); }
    public void SetException(Exception exception) { }
    public void SetResult() { }
    public void SetStateMachine(IAsyncStateMachine stateMachine) { }
    public void Start<TStateMachine>(ref TStateMachine stateMachine) where TStateMachine : IAsyncStateMachine { }
}

namespace System.Runtime.CompilerServices
{
    public class AsyncMethodBuilderAttribute : System.Attribute
    {
        public AsyncMethodBuilderAttribute(Type type) { }
    }
}
```

**Comment by jcouv** [comment]
This repros on latest master ([TryRoslyn](http://tryroslyn.azurewebsites.net/#b:master/K4Zwlgdg5gBAygTxAFwKYFsDcAoUlaIoYB0AKgBYBOqAhgCb5k0gDWIOe08Sa6xASsAjIw6VMQDCAe3QAHMABtUlOMoBuYAMap22bLOAAjBVpiaFzEDAAKlKVEo102AN7YYHmAeOmUNEZowalJgdDAAsjSQABQAlDAuMAC+2Cn6RiaB5pYwpAiyqNY0jmJolCC2Uhp0ygA8pAB8ru6e3pkwzAgQgQBqNArAqKTMLBGoyORSdHEtHm6eCx0A7lHIuSPEACKoFgjRAKyxOIue1MjAlBDHnmkLANoAgiBdmuHjk3QAQsCKNZTRyHyqCkADNok8Xn0BkMRm8JlNvr9lLFYgBdWZeDK+ZCUYCaNZQwbDVgY+Y3VJ6NqmbIgKwQ7qEmGsOEfREKP7NBZUwJ5ApFErjZQVOzVOqQZANYiM4mjGUJGBQcYwAC8DRgNRBNGACmQ0V5hWKTkF5UqosotXFkulIyOyQx3KCITCDxWYGQAHkINI5Eo0HR6i7VsoADS5ODIfyoSKaciQVANaLUEG5QNu5TLIOUUNJsMRtDR2MQVAwPz5mgxuPxJbkZTF0ipsowABcMAAkgA5KQiEEIb2yX1gKQQGDV2u5yMFuPNtv0zThiflwvFxIpLlYwLBUIwBsAVQgIBoINQnr7vtQ/vrrrKodI87LFaLCZzl8zGbTWZgz7vUcXU9LP4fVAqxragUyvdMW1bCRKDdLR+k7btexkftxkHYdR1A288wApdp1bWdv0nIt5VXVp1xLPNTFna1mXeBEfnZdNoNoNA4nlM4LmHDUtR1cFngZfoiVhOivgYv5bVIjwHU3MJVGQABRAAPbRZBEIdoiUlS1OHVBlNQVS0PiFd7XImT4HGfgdG1XUjLtNcfA3J1zOQQjfyLaJ8P4udsKI4t/182zJMxBzHS3edKGQepXMAp9UGTLCF0AijEqXYCxwS+9cMggifLc5c7TSCAjRAWRy2LQheAEIQRDESRkMUZRVEoDRtBATkyJCmk6S8ll6KRSgHmQHEwEMYA0GnCqSEG4bRrQUkMXs9pZ160T+ummDZtQPUgRgQECkCjEUiSIA===)).

**Comment by jcouv** [comment]
@cston, I couldn't find a feature doc for task-like, so I went back to the [feature proposal](https://github.com/dotnet/roslyn/issues/7169). From my understanding, `TypeParametersProvider<T>.ValueTask` cannot be "task-like" because it is neither (1) a concrete type nor (2) a generic type of arity 1.
Can you confirm?
If that is correct, I'm assuming we should just expect an error in the above scenario.

**Comment by jcouv** [comment]
It looks like my thinking was wrong. There are tests with task-like types nested in generic types, so this is supported.
Looking at this more, it looks like the problem is the mismatch between the return type of `Method` (returns `TypeParametersProvider<T>.ValueTask`) and the type created by the AsyncMethodBuilder (creates a `TypeParametersProvider<int>.ValueTask`).

**Comment by jcouv** [comment]
Tracked by VSO https://devdiv.visualstudio.com/DevDiv/_workitems?id=369807

**Comment by jcouv** [comment]
@TessenR Thanks for reporting this issue.
Fixed with #16530. Unfortunately, the fix won't make it into VS 2017 RTM. But it will ship in the first minor release after that.
